OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,421 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Adds an overlay for weather animations based on dropping particles (snowfalkes, raindrops, sand grains)")]
+	public class WeatherOverlayInfo : ITraitInfo
+	{
+		[Desc("On-off switch for the weather overlay")]
+		public readonly bool WeatherEnabled = true;
+		[Desc("Can be snow or rain or d2kSand")]
+		public readonly string WeatherVariants = "snow";
+		[Desc("Makes dropping particles not falling straight down")]
+		public readonly bool Wind = true;
+		[Desc("The wind is fading in from zero")]
+		public readonly bool InstantWind = false;
+		[Desc("The wind will change in intensity and direction over time")]
+		public readonly bool AutoWind = true;
+		[Desc("Sets the wind intensity manually and works only when AutoWind is disabled. Can be: StormLeft, StrongLeft, WeakLeft, Calm, WeakRight, StrongRight, StormRight")]
+		public readonly WeatherOverlay.WindTypes WindType = WeatherOverlay.WindTypes.Calm;
+
+		public object Create(ActorInitializer init) { return new WeatherOverlay(init.World, this); }
+	}
+
+	public class WeatherOverlay : ITick, IRenderWeatherOverlay
+	{
+
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+
+		public struct droppingParticle
+		{
+			public float PosX;
+			public float PosY;
+			public int   Size;
+			public float DirectionScatterX;
+			public float Gravity;
+			public float SwingOffset;
+			public float SwingSpeed;
+			public bool  SwingDirection;
+			public float SwingAmplitude;
+			public Color Color;
+			public Color TailColor;
+
+			public droppingParticle(float posX, float posY, int size, float directionScatterX, float gravity, float swingOffset, float swingSpeed, bool swingDirection, float swingAmplitude, Color color, Color tailColor)
+			{
+				PosX = posX;
+				PosY = posY;
+				Size = size;
+				DirectionScatterX = directionScatterX;
+				Gravity = gravity;
+				SwingOffset = swingOffset;
+				SwingSpeed = swingSpeed;
+				SwingDirection =  swingDirection;
+				SwingAmplitude = swingAmplitude;
+				Color = color;
+				TailColor = tailColor;
+			}
+		} 
+		public List<droppingParticle> DroppingParticleList = new List<droppingParticle>();
+		int MaxDrPaCount = CalculateDroppingParticleCount(Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
```

Comment:
"DrPa" isn't a good naming scheme here or in the other places. In most you can go for "Particles".